File: bfs/JunYoungKr/7562.py
from collections import deque
import sys
input = sys.stdin.readline

dx = [-2, -1, 1, 2, 2, 1, -1, -2]
dy = [-1, -2, -2, -1, 1, 2, 2, 1]

def bfs(x, y, len):
    queue = deque([(x, y)])
    
    while queue:
        x, y = queue.popleft()
        
        # 목표 위치에 도착하면 이동 횟수를 반환 (초기 값이 1이므로 -1 해줌)
        if x == fx and y == fy:
            return chess[x][y] - 1
        
        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if nx < 0 or nx >= len or ny < 0 or ny >= len:
                continue
            
            if chess[nx][ny] == 0:
                chess[nx][ny] = chess[x][y] + 1
                queue.append((nx, ny))
            
        
    
# 테스트 케이스 T 입력
T = int(input())

for _ in range(T):
    # 체스판의 한 변의 길이
    len = int(input())
    # 체스판 생성
    chess = [[0] * len for _ in range(len)]
    # 나이트가 현재 있는 칸
    x, y = map(int, input().split())
    # 나이트가 이동하려고 하는 칸
    fx, fy = map(int, input().split())
    
    if x == fx and y == fy:
        print(0)
        continue
    else:
        chess[x][y] = 1
        print(bfs(x, y, len))
        

File: BFS/DonggyunKim00/7562.js
const fs = require('fs');

const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt';
const input = fs
  .readFileSync(filePath, 'utf8')
  .toString()
  .trim()
  .split('\n')
  .map((item) => item.split(' ').map(Number));

const [dx, dy] = [
  [1, 2, 2, 1, -1, -2, -2, -1],
  [-2, -1, 1, 2, 2, 1, -1, -2],
];

const bfs = (length, start, end) => {
  let visited = Array.from({ length }, () => Array.from({ length }, () => -1));

  let curr = start;
  const queue = [curr];
  visited[curr[0]][curr[1]] = 0;

  while (queue.length) {
    const [prev_x, prev_y] = queue.shift();
    for (let dir = 0; dir < 8; dir++) {
      const mx = prev_x + dx[dir];
      const my = prev_y + dy[dir];

      if (mx < 0 || mx >= length || my < 0 || my >= length) continue;
      if (visited[mx][my] !== -1) continue;

      queue.push([mx, my]);
      curr = [mx, my];
      visited[mx][my] = visited[prev_x][prev_y] + 1;
    }
  }

  return visited[end[0]][end[1]];
};

const solution = (N, input) => {
  const answer = [];
  for (let i = 0; i < N; i++) {
    const [l, start, end] = input.slice(3 * i, 3 * i + 3);
    answer.push(bfs(l[0], start, end));
  }
  return answer;
};

const N = input.shift();
const answer = solution(Number(N), input).join('\n');
console.log(answer);
